agent: backhaul_manager: Implement AP Metrics Query/Response message

VladyslavTupikin opened the pull request:
Need to implement handlers for transmiting AP Metrics Query/Response messages
according to the UML diagram.

Add implementation for handlers for transmiting AP Metrics Query/Response messages.

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -653,12 +656,62 @@ bool monitor_thread::update_ap_stats()
+bool monitor_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
+                                                            ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto cmdu_message_type = cmdu_rx.getMessageType();
```
better put it after the next if statement, because the value is not used if the if is true.
The if statement is "escape" if.

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.h:
```diff
@@ -54,6 +54,9 @@ class monitor_thread : public beerocks::socket_thread {
     virtual std::string print_cmdu_types(const beerocks::message::sUdsHeader *cmdu_header) override;
 
 private:
+    bool handle_cmdu_control_ieee1905_1_message(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```
I noticed you escape this function if sd is nullptr. In this case, if you don't know how to work with a nullptr Socket, please _require_ it in your interface. The best would be:
```suggestion
    bool handle_cmdu_control_ieee1905_1_message(Socket &sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```

arnout replied:
I agree, but note that this is a shortcoming in this entire file. That said, for new functions we can definitely do the right thing.

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    const auto mid            = cmdu_rx.getMessageId();
+    auto ap_metrics_query_tlv = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
+    if (!ap_metrics_query_tlv) {
+        LOG(ERROR) << "AP Metrics Query CMDU mid=" << mid << " does not have AP Metric Query TLV";
```
Good! the mid is a hint you added for later investigation.

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+        LOG(ERROR) << "Failed building IEEE1905 AP_METRICS_QUERY_MESSAGE";
+        return false;
+    }
+    auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
```
These two lines are extremely weird. I guess they are OK because addCLass returns a pointer to internal member and then you set the value of this member. But please note that the structure of these two lines is like writing:
```cpp
A a;
auto x = a.x;
x = a.y;
```
which is strange.

RanRegev replied:
Also I don't understand why you named it "response".
This is not a response, you _forward_ the message.
```suggestion
    auto ap_metrics_query_tlv = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
```

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                   << "  bssid " << std::get<1>(bssid_tupl);
+    }
+
+    auto cmdu_header = cmdu_tx.create(0, ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE);
```
Why don't you use the same mid as received line 2326?
I mean like:
```suggestion
    auto cmdu_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE);
```

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
     return send_cmdu_to_bus(cmdu_tx, src_mac, bridge_info.mac);
 }
 
+bool backhaul_manager::handle_ap_metrics_query(ieee1905_1::CmduMessageRx &cmdu_rx,
```
Unless I'm missing something crucial - this function and the slave_thread function with the same name (6f9a662#diff-f3dcf05057f5ecc09b7bc39daa3fe7f2R4528) are doing the same:

- verification on the rx
- build tx from the data inthe rx.

Maybe it would be better to separate this functionality into one function that both threads would use?

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -653,12 +656,62 @@ bool monitor_thread::update_ap_stats()
-        LOG(ERROR) << "Not a vendor specific message";
-        return false;
+        LOG(DEBUG) << "Received IEEE 1905.1 message";
+        handle_cmdu_control_ieee1905_1_message(sd, cmdu_rx);
```
I've also had to implement this sort of differentiating and dispatching vendor-specific messages and IEEE 1905.1 messages. However, I've done it in a slightly different way, by checking the type of message. At this point we'll have a merge conflict and we should agree on which mechanism should be used.

arnout replied:
@mariomaz If you think your mechanism is better, please make a quick PR with only that commit, and Vlad can rebase on it.

mariomaz replied:
Done, see #1236.
New mechanism is also implemented in #1233

VladyslavTupikin replied:
So we do not have conflicts, @mariomaz please check updated changes

mariomaz replied:
My PRs have already been merged onto master. You should rebase your changes

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -653,12 +656,62 @@ bool monitor_thread::update_ap_stats()
+        LOG(ERROR) << "AP Metrics Query CMDU mid=" << mid << " does not have AP Metric Query TLV";
+        return false;
+    }
+    for (int bssid_idx = 0; bssid_idx < ap_metrics_query_tlv->bssid_list_length(); bssid_idx++) {
```
Type of bssid_idx should be `size_t`

arnout replied:
Well, bssid_list_length() is a uint8_t, so if anything, bssid_idx should be uint8_t...

arnout replied:
Never mind, size_t is better.

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -653,12 +656,62 @@ bool monitor_thread::update_ap_stats()
+        auto bssid = ap_metrics_query_tlv->bssid_list(bssid_idx);
+        if (!std::get<0>(bssid)) {
+            LOG(ERROR) << "Failed to get bssid " << bssid_idx << " from AP_METRICS_QUERY";
+            //return false;
```
This line should not be commented out.
You cannot use the second element of the tuple if the first one is false.

VladyslavTupikin replied:
outdated

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+        LOG(ERROR) << "AP Metrics Query CMDU mid=" << mid << " does not have AP Metric Query TLV";
+        return false;
+    }
+    for (int bssid_idx = 0; bssid_idx < ap_metrics_query_tlv->bssid_list_length(); bssid_idx++) {
```
Use `size_t`

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    }
+    auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
+    ap_metrics_response_tlv      = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
+    for (auto socket : slaves_sockets) {
```
I've also had to forward received message to slaves, but I did it in a different way using `message_com::forward_cmdu_to_uds`. which I think is more convenient than creating a new copy of the message.

VladyslavTupikin replied:
applied when forward required

mariomaz replied:
Why aren't you using `message_com::forward_cmdu_to_uds`?
I think code will be much simpler

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    ap_metrics_response_tlv      = ap_metrics_query_tlv;
+    LOG(DEBUG) << "Sending AP_METRICS_RESPONSE_MESSAGE, mid=" << std::hex << int(mid);
+
+    return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
```
I think that backhaul manager should wait until responses from all the slaves had been received. At that moment, backhaul manager should create and send a new response message with metrics received from all requested radios.

VladyslavTupikin replied:
Implemented this mechanism in the last update

RanRegev replied:
... However, In case of error, when not all responses were received, `ap_metric_query`  is left in an unstable state - it contains the metrics queries that did not receive any response.
This situation actually stops any farther queries from succeeding as the case to send a response is the emptiness of this vector (see above, line 2548)

arnout replied:
So I think a better approach would be to have a separate structure that contains:
- the list of slaves from which we expect a response;
- for each slave, the bssids for which we expect a response;
- the mid of the expected response (to distinguish from duplicate queries);
- the list of received response information.

When we receive a query, this structure is reset.

When we receive a response that is not expected, it is forwarded as-is to the controller.

This still leaves a weird situation when there are two requests back to back: since the "expected" of the first request is discarded, any responses that we receive are forwarded as is to the controller. But this is a situation that shouldn't happen, so any behaviour is fine as long as we don't crash and at least one of the requests is handled correctly.

mariomaz replied:
> When we receive a response that is not expected, it is forwarded as-is to the controller.

I'd ignore this one

mariomaz replied:
We also have to consider the case of periodically reporting AP metrics (without a query, no mid)

mariomaz commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4534,14 +4537,41 @@ bool slave_thread::handle_ap_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx
+        LOG(ERROR) << "Failed building IEEE1905 AP_METRICS_QUERY_MESSAGE";
+        return false;
+    }
+    message_com::send_cmdu(monitor_socket, cmdu_tx);
```
I've also used `message_com::forward_cmdu_to_uds` here to forward message to monitor thread.

mariomaz reviewed (changes requested):
We have used different approaches to implement similar things (see #1233). We should agree on which is the most appropriate way before continuing.

arnout commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -653,12 +656,62 @@ bool monitor_thread::update_ap_stats()
+        LOG(ERROR) << "Socket error";
+        return false;
+    }
+
```
(nitpick) We don't do such a check anywhere else in this file. Actually, it should really be passed by reference instead of as a pointer.

VladyslavTupikin replied:
We do it in monitor_thread::socket_disconnected

arnout replied:
That's different: that one may actually be called with a nullptr.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2092,6 +2094,10 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+    case ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE:
+        return handle_ap_metrics_query(cmdu_rx, src_mac);
+    case ieee1905_1::eMessageType::AP_METRICS_RESPONSE_MESSAGE:
+        return handle_ap_metrics_response(cmdu_rx, src_mac);
```
I don't think this is needed. This function is only called when the message comes from the bus, not when it comes from a slave. Instead, you have to add the handling to handle_slave_backhaul_message() (which at the moment is also only supporting Intel VS messages).

arnout reviewed (commented):
Not a real review, just pointing out where the bug is.

Commit:
agent: backhaul_manager: Implement AP Metrics Query/Response message

Need to implement handlers for transmiting AP Metrics Query/Response messages
according to the UML diagram.

Add implementation for handlers for transmiting AP Metrics Query/Response messages.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
tests: test_flow.py: Fix combined_infra_metrics test

Current implementation of combined_infra_metrics test sends
AP_METRIC_QUERY_MESSAGE with wrong data in tlv - radio mac instead
of the VAP bssid.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
tlvf: CmduMessageRx: parseNextTlv: add case for tlvApMetrics

In method parseNextTlv() in the file CmduMessageRx.cpp was missing a
case for processing tlvApMetrics. Add it.

Also add debug print to the log in the default case for avoiding long
debug session for finding the root case of nullptr after
cmdu_rx.getClass calls.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

RanRegev commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -219,6 +224,7 @@ std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv()
         return msg.addClass<wfa_map::tlvApCapability>();
     }
     default: {
+        LOG(DEBUG) << "tlvUnknown message type: " << tlv_type;
```
great.

mariomaz replied:
(nitpick) There is a "typo" in the log message: should be "Unknown" instead of "tlvUnknown".
Better off, "Unknown TLV type"

RanRegev commented on tests/test_flows.py:
```diff
@@ -493,51 +493,33 @@ def test_link_metric_query(self):
         debug("Send AP Metrics query message to agent 1")
         env.controller.dev_send_1905(env.agents[0].mac, 0x800B,
-                                     tlv(0x93, 0x0007, "0x01 {%s}" % (env.agents[0].radios[0].mac)))
+                                     tlv(0x93, 0x0007, "0x01 {%s}" % (env.agents[0].radios[0].vaps[0].bssid)))  # (env.agents[0].radios[0].vaps[0].bssid)))
```
The comment at the end of the line should be removed.

RanRegev commented on tests/test_flows.py:
```diff
@@ -493,51 +493,33 @@ def test_link_metric_query(self):
-        self.check_log(env.controller, "Received TLV_RECEIVER_LINK_METRIC")
+        self.check_log(env.controller, "Received AP_METRICS_RESPONSE_MESSAGE")
 
+        #debug("Send 1905 Link metric query to agent 1 (neighbor gateway)")
```
The final file seems to be ok: you send a request and looks for request and response in the log.
also you mentioned that the code is _temporary_ closed, which I guess means later the code would be removed.
this is fine. 
the only thing that I would do is add a comment above the commented line explaining that these lines should be removed in the future after \<this date\> / \<this event\> / \<before merge\> / etc.  Just for the case they will wrongly be left there.

mariomaz replied:
You'd better create an issue describing the work left to be done and copy the issue number in a TODO comment here

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -666,6 +670,87 @@ bool monitor_thread::update_ap_stats()
+{
+    const auto mid            = cmdu_rx.getMessageId();
+    auto ap_metrics_query_tlv = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
+    sMacAddr bssid_response;
```
Move after the escape `if` and before the `for` it is being used

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -666,6 +670,87 @@ bool monitor_thread::update_ap_stats()
+        LOG(ERROR) << "AP Metrics Query CMDU mid=" << mid << " does not have AP Metric Query TLV";
+        return false;
+    }
+    for (size_t bssid_idx = 0; bssid_idx < ap_metrics_query_tlv->bssid_list_length(); bssid_idx++) {
```
This seems like validation and log `for` loop. If so please comment above it accordingly.
Also the last line of the loop seems to be useless, the value in `bssid_response` after the loop would be of the _last_ in ap_metrics_query_tlv list. If this is the *desired* behavior it must be well documented as this is not natural.

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -666,6 +670,87 @@ bool monitor_thread::update_ap_stats()
+                   << "  bssid " << std::get<1>(bssid);
+        bssid_response = std::get<1>(bssid);
+    }
+    //TODO: add
```
confusing. not sure what is there is TODO. what should be added.

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -666,6 +670,87 @@ bool monitor_thread::update_ap_stats()
+    wfa_map::tlvApMetric::sEstimatedService estimated_service_parameters;
+    estimated_service_parameters.include_ac_be = 1;
+
+    ap_metrics_response_tlv->bssid()                               = bssid_response;
```
see comment above, bssid_response in the last in the list.
If this is ok, please comment the code correctly.

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1633,11 +1635,12 @@ bool backhaul_manager::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_r
             }
 
-            if (soc && soc->slave == sd) {
+            if (soc && soc->slave == sd &&
```
I still feel uncomfortable with this `if` statement, I don't have a suggestion, I know what I would like:
I want to be able to clearly understand the conditions that led me to a point in the code.
In the way it is written now, we get to `handle_slave_1905_1_message` if either is true:
* !soc
* !soc->slave
*  cmdu_rx.getMessageType() is not a VENDOR_SPECIFIC_MESSAGE

I don't have a suggestion.

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2118,13 +2121,29 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+        return handle_ap_metrics_response(cmdu_rx, src_mac);
+    }
+    default: {
+        LOG(DEBUG) << "Unknown 1905 message";
```
also would be great to add the message type for the error log.
```suggestion
        LOG(DEBUG) << "Unknown 1905 message:" << cmdu_rx.getMessageType()
```

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    auto bssid_tlv = ap_metric_tlv->bssid();
+    auto mac       = std::find_if(
+        ap_metric_query.begin(), ap_metric_query.end(),
+        [&](sApMetricsQuery const &ap_metric_query) { return ap_metric_query.bssid == bssid_tlv; });
```
It is better to be specific, just for clarity -
```suggestion
        [&bssid_tlv](sApMetricsQuery const &ap_metric_query) { return ap_metric_query.bssid == bssid_tlv; });
```
(However, I don't have strong opinion about this issue)

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+        return true;
+    }
+
+    // If all queries processed - prepare and send response message to the controller
```
```suggestion
    // at this point all queries processed - prepare and send response message to the controller
```

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    }
+
+    // If all queries processed - prepare and send response message to the controller
+    auto cmdu_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::AP_METRICS_RESPONSE_MESSAGE);
```
If I am not missing something, `mid` is the mid of the _last_ received response. Is this correct?

arnout replied:
They (should) have all the same MID. But in fact, this should be checked.

RanRegev reviewed (approved):
I have few considerations that need to be answered before merging

arnout commented on agent/src/beerocks/slave/son_slave_thread.h:
```diff
@@ -133,7 +133,7 @@ class slave_thread : public beerocks::socket_thread {
                                      std::shared_ptr<beerocks::beerocks_header> beerocks_header);
     bool handle_cmdu_control_ieee1905_1_message(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
-
+    bool handle_cmdu_monitor_ieee1905_1_message(Socket &sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```
Excellent naming!

arnout commented on agent/src/beerocks/slave/son_slave_thread.h:
```diff
@@ -238,7 +238,8 @@ class slave_thread : public beerocks::socket_thread {
     bool send_operating_channel_report();
-    bool handle_ap_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_ap_metrics_query(Socket &sd, ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_ap_metrics_response(Socket &sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```
Might be useful to include the fact that it comes from monitor in the name here...

arnout commented on tests/test_flows.py:
```diff
@@ -493,51 +493,33 @@ def test_link_metric_query(self):
-                                     tlv(0x93, 0x0007, "0x01 {%s}" % (env.agents[0].radios[0].mac)))
+                                     tlv(0x93, 0x0007, "0x01 {%s}" % (env.agents[0].radios[0].vaps[0].bssid)))  # (env.agents[0].radios[0].vaps[0].bssid)))
         self.check_log(env.agents[0].radios[0], "Received AP_METRICS_QUERY_MESSAGE")
         # TODO agent should send response autonomously, with same MID.
```
TODO should be removed.

Conversely, we should use the sniffer to verify that the agent sends the expected AP Metrics Response.

mariomaz replied:
This TODO comment should be removed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1633,11 +1635,12 @@ bool backhaul_manager::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_r
             } else {
-                LOG(ERROR) << "ACTION_BACKHAUL from none slave socket!";
-                return false;
+                return handle_slave_1905_1_message(cmdu_rx, src_mac);
```
This should also be inside the `if (soc && soc->slave == sd)` condition. The second branch of that condition is actually redundant, because the loop above sets `soc` to non-nullptr only if `soc->slave == sd`.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                                               const std::string &src_mac)
+{
+    std::vector<sMacAddr> bssid;
+    auto ret                  = false;
```
In the context of the "multiple points of exit" principle, something like this is suspicious. I'd rather expect patterns like:
```
if (!func()) {
   LOG(ERROR) << ...;
   return false;
}
```
So no `ret` is needed.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    auto ap_metrics_query_tlv = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
+    if (!ap_metrics_query_tlv) {
+        LOG(ERROR) << "AP Metrics Query CMDU mid=" << mid << " does not have AP Metric Query TLV";
+        return ret;
```
So here 
```suggestion
        return false;
```

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+        return ret;
+    }
+    for (size_t bssid_idx = 0; bssid_idx < ap_metrics_query_tlv->bssid_list_length(); bssid_idx++) {
+        auto bssid_tupl = ap_metrics_query_tlv->bssid_list(bssid_idx);
```
(nitpick) 
```suggestion
        auto bssid_tuple = ap_metrics_query_tlv->bssid_list(bssid_idx);
```

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -503,6 +508,28 @@ class backhaul_manager : public btl::transport_socket_thread {
+    } sApMetricsResponse;
+
+    size_t processed_queries = 0;
+    std::vector<sApMetricsQuery> ap_metric_query;
```
```suggestion
    std::vector<sApMetricsQuery> m_ap_metric_query;
```
(same for the others)

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    for (auto socket : slaves_sockets) {
+        for (auto mac : bssid) {
+            int i = 0;
+            if (mac == socket->vaps_list.vaps[i].mac) {
```
It would actually make more sense to have a structure like this:
```
for (size_t bssid_idx = 0; bssid_idx < ap_metrics_query_tlv->bssid_list_length(); bssid_idx++) {
    const auto &bssid = ...
    auto slave = find_slave_for_bssid(bssid);
    if (!slave) {
       // Error, send ACK with error code
    }
    if (slave not in m_ap_metric_query) {
        forward_cmdu
        add slave to m_ap_metric_query
    }
}
```

mariomaz replied:
m_ap_metric_query should be cleared before entering the loop, just in case there is data from a previous query message that has not been completely processed yet

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+            int i = 0;
+            if (mac == socket->vaps_list.vaps[i].mac) {
+                LOG(DEBUG) << "Forwarding AP_METRICS_QUERY_MESSAGE message to son_slave, bssid: "
+                           << std::hex << network_utils::mac_to_string(mac);
```
std::hex is useless here. So is the mac_to_string.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                std::get<0>(list) = true;
+                std::get<1>(list) = mac;
+
+                if (!message_com::send_cmdu(socket->slave, cmdu_tx)) {
```
Instead of forwarding a separate query with a single bssid, I think we should forward it as is.
```suggestion
                if (!message_com::forward_cmdu_to_slave(socket->slave, cmdu_rx, length)) {
```

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -666,6 +670,87 @@ bool monitor_thread::update_ap_stats()
+        return false;
+    }
+
+    auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetric>();
```
I think you should move this code out to a method. This way we'll be able to reuse it when channel utilization threshold is crossed.

arnout replied:
Since this will be used in multiple places, it would be better to create a new function for adding the tlvApMetrics.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -503,6 +508,28 @@ class backhaul_manager : public btl::transport_socket_thread {
+
+    size_t processed_queries = 0;
+    std::vector<sApMetricsQuery> ap_metric_query;
+    std::vector<sApMetricsResponse> ap_metric_response;
```
According to coding style rules, class data members should have an m_ prefix: m_processed_queries, m_ap_metric_query, m_ap_metric_response

VladyslavTupikin replied:
In the backhaul manager, we are searching socket for specific BSSID and we should send a query with concrete BSSID instead of forwarding all payload because in the monitor_thread we should fill up BSSID in the response message and we can't find the correct BSSID from the full payload as we did in the backhaul manager.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                if (!message_com::send_cmdu(socket->slave, cmdu_tx)) {
+                    LOG(ERROR) << "Failed forwarding AP_METRICS_QUERY_MESSAGE message to son_slave";
+                    ret = false;
+                    continue;
```
If received query message cannot be forwarded to one of the slaves, then this method should return false here.
And before that, probably rollback by cleaning ap_metric_query vector. This way, responses from slaves to which received query message has already been forwarded to will be ignored.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                } else {
+                    ret = true;
+                    // Fill a query vector
+                    ap_metric_query.push_back({socket->slave, mac});
```
You are not considering the case in which several AP_METRICS_QUERY_MESSAGE messages are received in a row, before their responses from slaves had been processed.
Maybe you should store the mid of the AP_METRICS_QUERY_MESSAGE to discard responses from slaves which belong to a previous request.

mariomaz reviewed (changes requested):
Request/response mechanism should be more robust

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    }
+
+    // Prepare tlvApMetric for each processed query
+    for (size_t i = 0; i < processed_queries; i++) {
```
We should just iterate over ap_metric_response here.
```suggestion
    for (response : ap_metric_response) {
```

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+            LOG(ERROR) << "Failed addClass<wfa_map::tlvApMetric>";
+            return false;
+        }
+        LOG(DEBUG) << "VT_back in tlv loop ";
```
I'm not sure what this is supposed to mean.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -503,6 +508,28 @@ class backhaul_manager : public btl::transport_socket_thread {
+        std::vector<uint8_t> estimated_service_info_field;
+    } sApMetric;
+    typedef struct sApMetricsResponse {
+        Socket *soc;
```
The soc field is not used for anything, so it can be removed.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -503,6 +508,28 @@ class backhaul_manager : public btl::transport_socket_thread {
+    } sApMetric;
+    typedef struct sApMetricsResponse {
+        Socket *soc;
+        sMacAddr bssid;
```
The bssid field is not used for anything, so it can be removed.

Bottom line: we don't need this structure, we can use metric directly.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2413,6 +2432,163 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                          "ap_metrics_response_tlv->alloc_estimated_service_info_field";
+            return false;
+        }
+        auto tmp = ap_metrics_response_tlv->estimated_service_info_field();
```
IMO this tmp variable is not helpful.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4553,6 +4568,38 @@ bool slave_thread::handle_ap_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx
+    if (!ap_metrics_response_tlv) {
+        LOG(ERROR) << "Failed cmdu_rx.getClass<wfa_map::tlvApMetric>()" << std::hex << int(mid);
+        return false;
+    }
```
Since we don't do anything with this, it's not needed to check it. Just forward it.

arnout commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -14,6 +14,10 @@
 #include <beerocks/tlvf/beerocks_message.h>
 
+#include <tlvf/ieee_1905_1/tlvMacAddress.h>
+#include <tlvf/wfa_map/tlvApMetric.h>
```
Hum, apparently the yaml file is not entirely accurate: it is called "AP Metrics TLV" so this should be `tlvApMetrics`.

arnout commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -14,6 +14,10 @@
 
+#include <tlvf/ieee_1905_1/tlvMacAddress.h>
+#include <tlvf/wfa_map/tlvApMetric.h>
+#include <tlvf/wfa_map/tlvApMetricQuery.h>
```
Note that in the query it *is* singular. Confusing spec...

arnout commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -666,6 +670,87 @@ bool monitor_thread::update_ap_stats()
+        }
+        LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE, mid=" << std::hex << int(mid)
+                   << "  bssid " << std::get<1>(bssid);
+        bssid_response = std::get<1>(bssid);
```
Here we should also check that the bssid is indeed one of our vaps. And all the rest below should be done inside the loop.

arnout commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -666,6 +670,87 @@ bool monitor_thread::update_ap_stats()
+        LOG(ERROR)
+            << "Couldn't allocate ap_metrics_response_tlv->alloc_estimated_service_info_field";
+    auto tmp = ap_metrics_response_tlv->estimated_service_info_field();
+    std::copy_n(buf, sizeof(buf), tmp);
```
Rather than copying from an uninitialized buffer, it would be better to use `fill_n` to set it to 0.

arnout reviewed (changes requested):
Lots of comments from various reviewers but as promised I'll take over this PR.

arnout committed:
tlvf: rename tlvApMetric to tlvApMetrics

The TLV in the query metric is called "AP Metric Query TLV", but in the
responsie it is "AP Metrics TLV". To preserve consistency, rename the
class to correspond to the standard.

Update all uses as well.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
First update done, but not all comments have been handled yet by far.

arnout committed:
agent: monitor_thread: pass Socket by reference

The Socket argument to the internal functions is passed by pointer, even
though it can never be nullptr. So we should use a reference instead.

The functions defined in the superclass (handle_cmdu,
socket_disconnected) remain unchanged.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

MuteSpirit commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -466,6 +467,30 @@ void monitor_stats::process()
+    auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetrics>();
+
+    if (!ap_metrics_response_tlv) {
+        LOG(ERROR) << "Couldn't addClass tlvApMetrics";
```
Please add 
```
return false;
```
after this line

MuteSpirit commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2431,164 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+        }
+        bssid.push_back(std::get<1>(bssid_tupl));
+        LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE, mid=" << std::hex << int(mid)
+                   << "  bssid " << std::get<1>(bssid_tupl);
```
(nitpicks) Double call of `std::get<1>(bssid_tupl)`
```suggestion
        const auto &bssid = std::get<1>(bssid_tuple);
        bssid.push_back(bssid);
        LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE, mid=" << std::hex << int(mid)
                   << "  bssid " << bssid;
```

arnout committed:
test_flows.py: remove manual AP metrics response

Now the AP metrics request/response flow has been implemented, remove
the manually generated response from the combined_infra_metrics test.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
Most comments are taken into account. Remaining will be done in separate PRs:
- test_flows -> #1288
- collecting responses (refactoring + make more robust): TODO

I think this one can and should be merged as is now.

mariomaz commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -343,7 +343,7 @@ bool node::link_metrics_data::add_receiver_link_metric(
 }
 
-bool node::ap_metrics_data::add_ap_metric_data(std::shared_ptr<wfa_map::tlvApMetric> ApMetricData)
+bool node::ap_metrics_data::add_ap_metric_data(std::shared_ptr<wfa_map::tlvApMetrics> ApMetricData)
```
(nitpick)
Maybe the name of the parameter should be fixed as well

arnout replied:
That would be a separate commit. Since this PR has already taken too long, I'm going to leave it for the database refactoring.

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1403,6 +1407,48 @@ bool monitor_thread::handle_multi_ap_policy_config_request(Socket &sd,
+bool monitor_thread::handle_ap_metrics_query(Socket &sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid            = cmdu_rx.getMessageId();
+    auto ap_metrics_query_tlv = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
```
Should be named ap_metric_query_tlv

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1403,6 +1407,48 @@ bool monitor_thread::handle_multi_ap_policy_config_request(Socket &sd,
+
+    auto cmdu_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::AP_METRICS_RESPONSE_MESSAGE);
+    if (!cmdu_header) {
+        LOG(ERROR) << "Failed building IEEE1905 AP_METRICS_RESPONSE_MESSAGE";
```
(nitpick) Message is not IEEE1905

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1403,6 +1407,48 @@ bool monitor_thread::handle_multi_ap_policy_config_request(Socket &sd,
+            return false;
+        }
+
+        //TODO: find the way how to check policy config was enabled or not
```
Policy config is stored in monitor_db, inside a member variable of type sApMetricsReportingInfo

arnout replied:
This is implemented in #1285; changing the TODO will just create a merge conflict, so I'm keeping it as is.

mariomaz commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -466,6 +467,31 @@ void monitor_stats::process()
 
+bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMacAddr &bssid)
+{
+    auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetrics>();
```
Name should be ap_metrics_tlv

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1403,6 +1407,48 @@ bool monitor_thread::handle_multi_ap_policy_config_request(Socket &sd,
+        const auto &bssid = std::get<1>(bssid_tuple);
+        LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE, mid=" << std::hex << int(mid)
+                   << "  bssid " << bssid;
+        if (!mon_stats.add_ap_metrics(cmdu_tx, bssid)) {
```
Maybe it should check if BSSID is valid before adding TLV

arnout replied:
Yes, that's part of the remaining refactoring.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2434,159 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+{
+    std::vector<sMacAddr> bssid;
+    const auto mid            = cmdu_rx.getMessageId();
+    auto ap_metrics_query_tlv = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
```
Name should be ap_metric_query_tlv

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2434,159 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+
+    auto ret = false;
+    for (auto socket : slaves_sockets) {
+        for (auto mac : bssid) {
```
const auto &

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2434,159 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                }
+
+                if (!query->alloc_bssid_list(1)) {
+                    LOG(ERROR) << "Failed allocate memory for bssid_list";
```
return false if error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2434,159 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+
+                auto query = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
+                if (!query) {
+                    LOG(ERROR) << "Failed addClass<wfa_map::tlvApMetricQuery>";
```
return false if error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2434,159 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+                auto forward =
+                    cmdu_tx.create(mid, ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE);
+                if (!forward) {
+                    LOG(ERROR) << "Failed to create AP_METRICS_QUERY_MESSAGE";
```
return false if error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2434,159 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received AP_METRICS_RESPONSE_MESSAGE, mid=" << std::hex << int(mid);
+
+    auto ap_metric_tlv = cmdu_rx.getClass<wfa_map::tlvApMetrics>();
```
Name should be ap_metrics_tlv

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2412,6 +2434,159 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
+
+    // Prepare tlvApMetrics for each processed query
+    for (const auto &response : m_ap_metric_response) {
+        auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetrics>();
```
Name should be ap_metrics_tlv

arnout replied:
I disagree, I think it's good to distinguish from the ap_metrics_tlv variable that is used above.

mariomaz reviewed (approved):
A few nitpick comments and some others that I don't know if you want to address in this PR or in a coming PR, so I've removed the "ready for merge" label and approve to let you decide.
